tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,767 @@
+	bytesWritten += n
+
+	if err := cdm.writeAndAppendToCRC32(cdm.writeBuf[:bytesWritten]); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeAndAppendToCRC32(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeCRC32(); err != nil {
+		return 0, err
+	}
+
+	if maxt > cdm.curFileMaxt {
+		cdm.curFileMaxt = maxt
+	}
+
+	if mint < cdm.curFileMint {
+		cdm.curFileMint = mint
+	}
+
+	cdm.chunkBuffer.put(chkRef, chk)
+
+	if len(chk.Bytes())+MaxHeadChunkMetaSize >= writeBufferSize {
+		// The chunk was bigger than the buffer itself.
+		// Flushing to not keep partial chunks in buffer.
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	return chkRef, nil
+}
+
+func chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+// shouldCutNewFile decides the cutting of a new file based on time and size retention.
+// Size retention: because depending on the system architecture, there is a limit on how big of a file we can m-map.
+// Time retention: so that we can delete old chunks with some time guarantee in low load environments.
+func (cdm *ChunkDiskMapper) shouldCutNewFile(chunkSize int, maxt int64) bool {
+	return cdm.curFileNumBytes == 0 || // First head chunk file.
+		(maxt-cdm.curFileMint > cdm.maxFileMs && cdm.curFileNumBytes > HeadChunkFileHeaderSize) || // Time duration reached for the existing file.
+		cdm.curFileNumBytes+int64(chunkSize+MaxHeadChunkMetaSize) > MaxHeadChunkFileSize // Exceeds the max head chunk file size.
+}
+
+func (cdm *ChunkDiskMapper) cut(mint int64) (returnErr error) {
+	// Sync current tail to disk and close.
+	if err := cdm.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(cdm.dir, MagicHeadChunks, headChunksFormatV1, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// The file should not be closed if there is no error,
+		// its kept open in the ChunkDiskMapper.
+		if returnErr != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(returnErr)
+			merr.Add(f.Close())
+			returnErr = merr.Err()
+		}
+	}()
+
+	cdm.size += cdm.curFileNumBytes
+	atomic.StoreInt64(&cdm.curFileNumBytes, int64(n))
+
+	if cdm.curFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(cdm.curFile.Name())
+		if err != nil {
+			return err
+		}
+		cdm.readPathMtx.Lock()
+		oldFileCloser := cdm.closers[cdm.curFileSequence]
+		cdm.closers[cdm.curFileSequence] = newTailFile
+		cdm.mmappedChunkFiles[cdm.curFileSequence] = &mmappedChunkFile{
+			byteSlice: realByteSlice(newTailFile.Bytes()),
+			maxt:      cdm.curFileMaxt,
+		}
+		cdm.readPathMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldFileCloser.Close(); err != nil {
+			return err
+		}
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxHeadChunkFileSize))
+	if err != nil {
+		return err
+	}
+
+	cdm.curFileSequence = seq
+	cdm.curFileMint = mint
+	cdm.curFile = f
+	if cdm.wbuf != nil {
```

Comment:
It will require a new file to be cut in `newChunkDiskMapper` for this to be set, which I don't suggest (as it can lead to many empty files). A similar logic was used in writing blocks to avoid having any empty files.

Reply:
Yeo I see what you mean